File: API/models/Background.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field
from models.Generico import BaseSchema, ResourceReference


class EquipmentItemSchema(BaseModel):
    equipment: ResourceReference
    quantity: int

    model_config = ConfigDict(extra="allow")


class FeatureSchema(BaseModel):
    name: str
    desc: List[str] = Field(default_factory=list)

    model_config = ConfigDict(extra="allow")


class BackgroundSchema(BaseSchema):
    index: Optional[str] = None
    name: str
    starting_proficiencies: List[ResourceReference] = Field(default_factory=list)
    language_options: Optional[Dict[str, Any]] = None
    starting_equipment: List[EquipmentItemSchema] = Field(default_factory=list)
    starting_equipment_options: List[Dict[str, Any]] = Field(default_factory=list)
    feature: Optional[FeatureSchema] = None
    personality_traits: Optional[Dict[str, Any]] = None
    ideals: Optional[Dict[str, Any]] = None
    bonds: Optional[Dict[str, Any]] = None
    flaws: Optional[Dict[str, Any]] = None
    url: Optional[str] = None
    meta: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(extra='allow')

File: APP/src/app/interfaces/background.ts
export interface Background {
  id?: string;
  index: string;
  name: string;
  image?: string;
  starting_proficiencies?: BackgroundReference[];
  starting_equipment?: BackgroundEquipment[];
  starting_equipment_options?: Array<{ desc?: string }>;
  // Multiple proficiency choice groups (e.g., "choose 1 skill from {list}, choose 1 gaming set from {list}")
  starting_proficiency_options?: Array<{
    choose: number;
    type: string;
    options: string[];
  }>;
  feature?: {
    name: string;
    desc?: string[];
    is_variant?: boolean;
    variant?: {
      name: string;
      desc?: string[];
    };
  };
  // API estructura: { from: { options: [{ option_type: 'string', string: '...' }] }
  personality_traits?: BackgroundOptionGroup;
  ideals?: BackgroundOptionGroup;
  bonds?: BackgroundOptionGroup;
  flaws?: BackgroundOptionGroup;
  created_by?: string;
  created_at?: string;
  updated_at?: string;
  [key: string]: unknown;
}

export interface BackgroundOptionGroup {
  choose?: number;
  type?: string;
  // API format: { from: { options: [...] } }
  from?: {
    option_set_type?: string;
    options?: Array<{
      option_type?: string;
      string?: string;
      desc?: string;
      alignments?: Array<{
        index: string;
        name: string;
        url: string;
      }>;
    }>;
  };
  // Legacy form format: { options: [...] }
  options?: string[];
}

export interface BackgroundReference {
  index: string;
  name: string;
  url: string;
}

export interface BackgroundEquipment {
  equipment: BackgroundReference;
  quantity: number;
}